### ISSUE TITLE
Cannot read property 'values' of undefined

### ISSUE THREAD

**Issue by pgrekovich** [jaredpalmer/formik-effect#4]
One of these days there was this error. Any idea what might be the problem?

![screenshot_3](https://user-images.githubusercontent.com/1001687/40150760-555c0ba8-5984-11e8-97df-c92251c7adcc.png)

**Comment by latviancoder** [comment]
formik is using new Provider/Consumer from react 16.3, while this library still relies on old context api.

@jaredpalmer are you planning on updating this?

**Comment by maticrivo** [comment]
@jaredpalmer any update or plans on this?

**Comment by pgrekovich** [comment]
@maticrivo I had spent time to understand it, and in real life, you don't need it!  Try to use componentDidUpdate if you need to update something after the form has updated.

**Comment by maticrivo** [comment]
@pgrekovich `componentDidUpdate` will only work if I use formik as an HOC with `withFormik` but i use formik with a `render` prop

**Comment by NamPNQ** [comment]
Just use connect api
Here is my solution
```jsx
// @flow
import React from 'react';
import { connect } from 'formik';

type Props = {
  onChange: Function,
  formik: Object,
};

class Effect extends React.Component<Props> {
  componentWillReceiveProps(nextProps) {
    const { values, touched, errors, isSubmitting } = this.props.formik;
    const {
      values: nextValues,
      touched: nextTouched,
      errors: nextErrors,
      isSubmitting: nextIsSubmitting,
    } = nextProps.formik;
    if (nextProps.formik !== this.props.formik) {
      this.props.onChange(
        {
          values,
          touched,
          errors,
          isSubmitting,
        },
        {
          values: nextValues,
          touched: nextTouched,
          errors: nextErrors,
          isSubmitting: nextIsSubmitting,
        },
      );
    }
  }

  // eslint-disable-next-line
  render() {
    return null;
  }
}

export default connect(Effect);
```

**Comment by sreeram-muddu** [comment]
@jaredpalmer any update on this issue?

**Comment by luke-dare** [comment]
Hey .. I've got the same issue.

**Comment by h0jeZvgoxFepBQ2C** [comment]
@jaredpalmer ping, same issue here

**Comment by brandonhall** [comment]
My solution was very similar to @NamPNQ and works great for us:

```typescript
import { Component } from 'react';
import { debounce, isEqual } from 'lodash';
import { connect, FormikValues, FormikProps } from 'formik';

interface IProps {
  onChange: (values: FormikValues) => void;
}

interface IPropsEnhanced extends IProps {
  formik: FormikProps<FormikValues>;
}

const SAVE_DELAY = 1000;

class FormikEffects extends Component<IPropsEnhanced> {
  onChange = debounce(this.props.onChange, SAVE_DELAY);

  componentDidUpdate(prevProps: IPropsEnhanced) {
    const { formik } = this.props;
    const { isValid } = formik;

    const hasChanged = !isEqual(prevProps.formik.values, formik.values);
    const shouldCallback = isValid && hasChanged;

    if (shouldCallback) {
      this.onChange(formik.values);
    }
  }

  render() {
    return null as null;
  }
}

export default connect<IProps>(FormikEffects);
```

**Comment by Ridder90** [comment]
@brandonhall  Thanks for sharing! Using your example gives me `IProps is not defined`.  Did you define it somewhere else?

**Comment by brandonhall** [comment]
@Ridder90 No, it's defined at the top of the file as a TypeScript interface.

**Comment by Ridder90** [comment]
Hmmm strange...running it with `export default connect(FormikEffects);` works perfectly though.

**Comment by sebastianpatten** [comment]
Any eta when this will be fixed? I'm getting the same issue

**Comment by BrianBusby** [comment]
Same here; would love a fix for this

**Comment by aronkof** [comment]
> Just use connect api
> Here is my solution
> 
> ```js-jsx
> // @flow
> import React from 'react';
> import { connect } from 'formik';
> 
> type Props = {
>   onChange: Function,
>   formik: Object,
> };
> 
> class Effect extends React.Component<Props> {
>   componentWillReceiveProps(nextProps) {
>     const { values, touched, errors, isSubmitting } = this.props.formik;
>     const {
>       values: nextValues,
>       touched: nextTouched,
>       errors: nextErrors,
>       isSubmitting: nextIsSubmitting,
>     } = nextProps.formik;
>     if (nextProps.formik !== this.props.formik) {
>       this.props.onChange(
>         {
>           values,
>           touched,
>           errors,
>           isSubmitting,
>         },
>         {
>           values: nextValues,
>           touched: nextTouched,
>           errors: nextErrors,
>           isSubmitting: nextIsSubmitting,
>         },
>       );
>     }
>   }
> 
>   // eslint-disable-next-line
>   render() {
>     return null;
>   }
> }
> 
> export default connect(Effect);
> ```

Thanks, it works as expected :)

**Comment by leechingching** [comment]
> My solution was very similar to @NamPNQ and works great for us:
> 
> ```ts
> import { Component } from 'react';
> import { debounce, isEqual } from 'lodash';
> import { connect, FormikValues, FormikProps } from 'formik';
> 
> interface IProps {
>   onChange: (values: FormikValues) => void;
> }
> 
> interface IPropsEnhanced extends IProps {
>   formik: FormikProps<FormikValues>;
> }
> 
> const SAVE_DELAY = 1000;
> 
> class FormikEffects extends Component<IPropsEnhanced> {
>   onChange = debounce(this.props.onChange, SAVE_DELAY);
> 
>   componentDidUpdate(prevProps: IPropsEnhanced) {
>     const { formik } = this.props;
>     const { isValid } = formik;
> 
>     const hasChanged = !isEqual(prevProps.formik.values, formik.values);
>     const shouldCallback = isValid && hasChanged;
> 
>     if (shouldCallback) {
>       this.onChange(formik.values);
>     }
>   }
> 
>   render() {
>     return null as null;
>   }
> }
> 
> export default connect<IProps>(FormikEffects);
> ```

It works as expected, but the console shows the warming below 
```
react-dom.development.js:506 Warning: A component is changing an uncontrolled input of type undefined to be controlled. Input elements should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled input element for the lifetime of the component. More info: https://fb.me/react-controlled-components
```

Do you have any ideas how to solve it ?

**Comment by NamPNQ** [comment]
@leechingching it's not related with this issue, you can check the guide in https://fb.me/react-controlled-components, may be your component dont set value back in input

**Comment by jariz** [comment]
Here's my version recreating formik-effect with hooks in 11 lines.  
[See here for TS version](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=25&pc=1#code/JYWwDg9gTgLgBAbzgYwgOzQU2TANHAMWhGAGsBhdGTAD3gF84AzKCEOAcieLI4G4AUKEixEcAEqYAhjgCiAG0whMaPHACuAZ0yymTbGq2ZJTOIxZtOUaTn4CBAegdwQATzhSY1cPE9xryNaewGgA5sw8pAC0mHoGcAAU4gCSAApwgA1kgPB-AJT2IdRQTDKYcADy6oWprGCaADwAalLy6piaAHyIAnA9cOjkABZSYZgAXInI6lBQAMownmOEkZSqtDCNza0d+GDWAG5zC+NEUCQUVGsbLW2dAD5waOry8jlwALydexDAACaC9PlVkUSnBqhBalctp01iofppypVMFAwRCmtcOl1ehFTmRjssLnRITd-vZUGhNPBdPocO84ESOgkkNwcaR8P0hiMzOMUfU0VDXh9Mb0yRT-LFaUYTHUTmcVtRCXybnB7o9nu0EqqXoIsUYqQYEgkBZ0EN0sb12cNQpgEsyzvhrEwAHSTaYqGA5bVmnoO51TayqWm2sienr0fAAbSDpAAuh77FjrDApmgHk95P9BAJaCJ4D9YlInvAyVgcAk9ThXlJ4fT1btwZpxhUqjVeZsbkaJDYYAolG7BEA).

```js
import { connect } from 'formik';
import { useEffect, useRef } from 'react';
const Effect = ({ formik, onChange }) => {
    const ref = useRef(null);
    useEffect(() => {
        onChange(formik, ref.current);
        ref.current = formik;
    }, [formik]);
    return null;
};
export default connect(Effect);

```
